### PR TITLE
HS - feat: optimize "properties" endpoint

### PR DESCRIPTION
**Pull request by akashrpo**
## Description of the change

> Optimize the call for "properties" endpoint to avoid reaching its rate-limit.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/hs/transform.js [hunk 1]
```diff
@@ -324,6 +326,8 @@ const processRouterDest = async inputs => {
   const errorRespList = [];
   // using the first destination config for transforming the batch
   const { destination } = inputs[0];
+  // reduce the no. of calls for properties endpoint
+  const propertyMap = await getProperties(destination);
```
Comment (saikumarrs) [inline v0/destinations/hs/transform.js:330]: Properties are needed if at least one of the events has `traits` defined. So, let's further optimize it.

File: v0/destinations/hs/transform.js [hunk 2]
```diff
@@ -218,8 +217,11 @@ async function processSingleMessage(message, destination) {
...
 }
 
-function process(event) {
-  return processSingleMessage(event.message, event.destination);
+// has been deprecated - using routerTransform for both the versions
+async function process(event) {
+  // no. of calls for properties endpoint remains the same
+  const propertyMap = await getProperties(event.destination);
+  return processSingleMessage(event.message, event.destination, propertyMap);
 }
```
Comment (saikumarrs) [inline v0/destinations/hs/transform.js:225]: Although not really a bad change, I would revert these changes and just send `undefined` in place of `propertyMap` to align with my comment below.

File: v0/destinations/hs/transform.js [hunk 3]
```diff
@@ -76,7 +76,6 @@ async function getTransformedJSON(message, mappingJson, destination) {
 
   if (traits) {
     const traitsKeys = Object.keys(traits);
-    const propertyMap = await getProperties(destination);
```
Comment (saikumarrs) [inline v0/destinations/hs/transform.js:79]: I would still keep the `destination` argument and calculate `propertyMap` if the incoming value is not defined.

Comment (akashrpo) [reply v0/destinations/hs/transform.js:79]: added the check

File: v0/destinations/hs/transform.js (hunk 2 above)
Comment (akashrpo) [reply v0/destinations/hs/transform.js]: sent undefined. The code looks cleaner this way.

File: v0/destinations/hs/transform.js (hunk 1 above)
Comment (akashrpo) [reply v0/destinations/hs/transform.js]: added the `traits` check

File: v0/destinations/hs/transform.js [hunk 4]
```diff
@@ -324,6 +334,21 @@ const processRouterDest = async inputs => {
...
+  let propertyMap;
+  const traitsFound = inputs.some(input => {
+    if (
+      input.message.traits ||
+      input.message.context.traits ||
+      input.message.properties
+    ) {
+      return true;
+    }
+    return false;
```
Comment (saikumarrs) [inline v0/destinations/hs/transform.js:347]: The logic above is not properly replicated here. Let's refactor that original logic into a method and use it in two places.

Comment (akashrpo) [reply v0/destinations/hs/transform.js:347]: The above original logic extracts the traits and here I think we just need to add the check for the presence of traits.

Comment (saikumarrs) [reply v0/destinations/hs/transform.js:347]: I can think of two reasons here for the re-factoring:
- Extracting traits (in general) is not a costly operation.
- Having this logic in two places is a potential risk going forward.

Comment (akashrpo) [reply v0/destinations/hs/transform.js]: refactored the logic into a method and used it in two different places.

File: v0/destinations/hs/transform.js [hunk 5]
```diff
@@ -324,6 +340,17 @@ const processRouterDest = async inputs => {
...
   // using the first destination config for transforming the batch
   const { destination } = inputs[0];
+  // reduce the no. of calls for properties endpoint
+  let propertyMap;
+  const traitsFound = inputs.some(input => {
+    if (getTraits(input.message)) {
+      return true;
+    }
+    return false;
+  });
```
Comment (saikumarrs) [inline v0/destinations/hs/transform.js:350]: Let's simplify this logic a bit,

```suggestion
  const traitsFound = inputs.some(input => {
    getTraits(input.message) !== undefined
    }
  });
```

Comment (akashrpo) [reply v0/destinations/hs/transform.js]: simplified

**Review by saikumarrs — APPROVED** [review]
